### PR TITLE
Adds plugin 'synchronize to device' #480

### PR DESCRIPTION
**Pull request by Tamriel**
Described in #480.
screenshot_20171125_005019

### CODE REVIEW

File: quodlibet/quodlibet/ext/events/synchronize_to_device.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        "Needs the python package 'pydub'.")
+    c_path = __name__ + '_path'
+
+    def PluginPreferences(self, parent):
+        vbox = Gtk.VBox(spacing=6)
+
+        queries = {}
+        query_path = os.path.join(get_user_dir(), 'lists', 'queries.saved')
+        try:
+            with open(query_path, 'rU') as query_file:
```
Comment (lazka) [inline quodlibet/quodlibet/ext/events/synchronize_to_device.py]: open(query_path, 'r', encoding="utf-8")

File: quodlibet/quodlibet/ext/events/synchronize_to_device.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+                log = Gtk.TextView()
+                log.set_left_margin(5)
+                log.set_right_margin(5)
+                log.editable = False
+                buffer = log.get_buffer()
+                scroll = Gtk.ScrolledWindow()
+                scroll.set_size_request(-1, 100)
+                scroll.add(log)
+
+                def append(text):
```
Comment (lazka) [inline quodlibet/quodlibet/ext/events/synchronize_to_device.py]: This is racy, the buffer and log could be destroyed one idle_add calls the callback.

File: quodlibet/quodlibet/ext/events/synchronize_to_device.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                for query_string in query_file:
+                    name = next(query_file).strip()
+                    queries[name] = Query(query_string.strip())
+
+                for query_name, query in queries.items():
+                    check_button = ConfigCheckButton(query_name, "plugins", self._config_key(query_name))
+                    check_button.set_active(self.config_get_bool(query_name))
+                    vbox.pack_start(check_button, False, True, 0)
+
+                def synchronize():
```
Comment (lazka) [inline quodlibet/quodlibet/ext/events/synchronize_to_device.py]: All the quodlibet API isn't thread safe (except if explicitly stated), so threads wont work here.

File: quodlibet/quodlibet/ext/events/synchronize_to_device.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                def stop(_):
+                    append("Stopping...")
+                    self.running = False
+
+                def path_changed(entry):
+                    self.path = entry.get_text()
+                    config.set('plugins', self.c_path, self.path)
+
+                try:
+                    self.pattern = config.get('plugins', self.c_path)
```
Comment (lazka) [inline quodlibet/quodlibet/ext/events/synchronize_to_device.py]: get() can be passed a default value

File: quodlibet/docs/packaging.rst [hunk 5]
```diff
@@ -117,6 +117,8 @@ plugin from loading.
 **libappindicator-gtk3** + **typelibs**:
     * For the Tray Icon plugin under Ubuntu Unity and KDE Plasma
 
+**pydub**:
```
Comment (lazka) [inline quodlibet/docs/packaging.rst]: (It's a bit unfortunate that pydub is needed, since theoretically we already ship everything for transcoding with GStreamer, but we can always change it at some later point)

File: quodlibet/quodlibet/ext/events/synchronize_to_device.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+                vbox.pack_start(destination_path_box, True, True, 0)
+                vbox.pack_start(start_button, True, True, 0)
+                vbox.pack_start(stop_button, True, True, 0)
+                vbox.pack_start(Gtk.Label(label=_("Output:")), True, True, 0)
+                vbox.pack_start(scroll, True, True, 0)
+                return qltk.Frame(_("The following saved searches shall be synchronized:"), child=vbox)
+        except FileNotFoundError:
+            return qltk.Frame(_("No saved searches yet, create some and come back!"))
+        except Exception as e:
```
Comment (lazka) [inline quodlibet/quodlibet/ext/events/synchronize_to_device.py]: Let uncaught exception through, so we get a crash report and can fix them.

File: quodlibet/quodlibet/ext/events/synchronize_to_device.py (hunk 3 above)
Comment (Tamriel) [reply quodlibet/quodlibet/ext/events/synchronize_to_device.py]: Thanks for the excellent code review!
Without using threads the UI will become unresponsive until the synchronisation has finished.
What could happen due to the not-thread-safe API? I think the worst scenario are wrong songs on the device.
The user is aware of the ongoing sync and won't edit the library.
So I propose to keep the threads.

Comment (Tamriel) [reply quodlibet/quodlibet/ext/events/synchronize_to_device.py]: So what shall we do here?

**Comment by Tamriel** [comment]
> All the quodlibet API isn't thread safe (except if explicitly stated), so threads wont work here.

But without threads, we can't show a process bar and the UI will become unresponsive, so we should keep threads.
We have two options now:

a) Use a modal dialog and show the sync process there. The dialog blocks user interaction with the library.

b) Allow using and editing the library while syncing, because the sync may take an hour and the user still wants to control what music he listens to.
 If the user adds songs to the library while syncing, then he won't have the new songs on his device. I think this is natural behaviour, the user is aware of this and will just re-sync after editing the library.

As soon as we have a decision I will polish this PR.

@frestr What do you think?

**Comment by frestr** [comment]
> But without threads, we can't show a process bar and the UI will become unresponsive, so we should keep threads.

I haven't tested or looked thoroughly through this PR yet, but a common way to do this in gtk applications is to have the following snippet run regularly:

```python
while Gtk.events_pending():
    Gtk.main_iteration()
```

This ensures gtk can process user events and prevents the application from becoming unreponsive. It might be possible to have someting like that between each song synced?

> We have two options now:

The second option makes the most sense to me, given that the above suggestion works.

**Comment by frestr** [comment]
You might also want to take a look at how the "Export Playlist to Folder" plugin is implemented. It uses the copool to do copying and regularly yields to let the UI update. It makes the UI pretty slow, but it's still usable to some extent.

File: quodlibet/quodlibet/ext/events/synchronize_to_device.py (hunk 2 above)
Comment (Tamriel) [reply quodlibet/quodlibet/ext/events/synchronize_to_device.py]: How could the buffer and log be destroyed?

**Comment by Tamriel** [comment]
I have some problems with your language tests. How can I fix them?

**Comment by Tamriel** [comment]
The test complains my plugin name 'Synchronize To Device' should be in title case like  'Last.fm Cover Source', 'Last.fm Sync', 'Send to iFP', 'This is a test'.
I think it is, so the test may be wrong?

**Comment by frestr** [comment]
That test is perhaps a bit too strict (with not very helpful error messages). The reason it fails here is that you have written "To" with a capital T. "Synchronize to Device" works.

**Comment by Tamriel** [comment]
Thanks for your help.
The only thing left is the maybe racy stuff [above](https://github.com/quodlibet/quodlibet/pull/2636#discussion-diff-159198577R40).
Is that important?

File: quodlibet/quodlibet/ext/events/synchronize_to_device.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+                            # prevent the application from becoming unreponsive
+                            while Gtk.events_pending():
+                                Gtk.main_iteration()
+
+                            song_path = song['~filename']
+                            song_file_name = "{} - {} - {} - {}.mp3".format(
+                                song("genre"), song("artist"),
+                                song("album"), song("title"))
+                            valid_chars = "-_.() %s%s" % (
```
Comment (frestr) [inline quodlibet/quodlibet/ext/events/synchronize_to_device.py]: Why is this here? Unicode characters should be allowed. This also generates duplicate filenames if the only difference in the filename is unicode characters.

File: quodlibet/quodlibet/ext/events/synchronize_to_device.py (hunk 2 above)
Comment (frestr) [reply quodlibet/quodlibet/ext/events/synchronize_to_device.py]: Do you need to use idle_add here?

File: quodlibet/quodlibet/ext/events/synchronize_to_device.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+                        append("Synchronization finished.")
+                    except FileNotFoundError as e:
+                        append(str(e))
+
+                def start(_):
+                    append("Starting...")
+                    self.running = True
+                    synchronize()
+
+                def stop(_):
```
Comment (frestr) [inline quodlibet/quodlibet/ext/events/synchronize_to_device.py]: Use `button` or something similar as the first argument instead, as this overrides `_` which we use for translation.

File: quodlibet/quodlibet/ext/events/synchronize_to_device.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+                vbox.pack_start(start_button, True, True, 0)
+                vbox.pack_start(stop_button, True, True, 0)
+                vbox.pack_start(Gtk.Label(label=_("Progress:")), True, True, 0)
+                vbox.pack_start(scroll, True, True, 0)
+                return qltk.Frame(
+                    _("The following saved searches shall be synchronized:"),
+                    child=vbox)
+        except FileNotFoundError:
+            return qltk.Frame(
+                _("No saved searches yet, create some and come back!"))
```
Comment (frestr) [inline quodlibet/quodlibet/ext/events/synchronize_to_device.py]: Also check if the file is empty.

File: quodlibet/quodlibet/ext/events/synchronize_to_device.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+                            enabled_queries.append(query)
+
+                    selected_songs = []
+                    for song in app.library.itervalues():
+                        if any(query.search(song) for query in
+                               enabled_queries):
+                            selected_songs.append(song)
+
+                    filename_list = []
+                    destination_path = config.get('plugins', self.c_path)
```
[the remark this answers is not in the document]
Comment (frestr) [reply quodlibet/quodlibet/ext/events/synchronize_to_device.py]: By default value I mean any value, such that no exception is thrown. The empty string `''` would be good here.

File: quodlibet/quodlibet/ext/events/synchronize_to_device.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+                    self.running = False
+
+                def path_changed(entry):
+                    self.path = entry.get_text()
+                    config.set('plugins', self.c_path, self.path)
+
+                self.pattern = config.get('plugins', self.c_path,
+                                          'the path of your device, e.g. '
+                                          '/run/media/my-username/device-id/'
+                                          'Music')
```
[the remark this answers is not in the document]
Comment (frestr) [reply quodlibet/quodlibet/ext/events/synchronize_to_device.py]: You can set a tooltip on the entry with `destination_path.set_tooltip_text()`.

File: quodlibet/quodlibet/ext/events/synchronize_to_device.py (hunk 7 above)
Comment (frestr) [reply quodlibet/quodlibet/ext/events/synchronize_to_device.py]: That definitely exists if you think about non-latin scripts like the ones used in Japanese, Chinese etc. Ideally there would be some fallback or option for this, but it should be okay for now.

**Comment by frestr** [comment]
> Is something wrong with the tests?

Yeah, you need to rebase to include 2926f2422924f0ebd30cd52c87e859df6b2680df.

> Regarding idle_add: It's not a method of the scolled object

What do you mean by this?

**Comment by Tamriel** [comment]
Is the test config broken?

**Comment by frestr** [comment]
> One test fails.
Is the test config of travis broken?

It just timed out. Fixed by a rerun.

**Comment by frestr** [comment]
So it looks alright now, but is there a reason you can't have append like this:

```python
def append(text):
    buffer.insert(buffer.get_end_iter(), text + "\n")
    log.scroll_to_mark(buffer.get_insert(), 0.0,
                       True, 0.5, 0.5)
```

without the idle_add?

**Comment by Tamriel** [comment]
> So it looks alright now, but is there a reason you can't have append like this
It works your way, so I added it. I guess the test timed out again?

**Comment by frestr** [comment]
> I guess the test timed out again?

You have syntax errors in your file.

**Comment by Tamriel** [comment]
I have no idea why that test fails.

**Comment by Tamriel** [comment]
It is ready to be merged :-)

**Comment by declension** [comment]
@lazka are you happy to merge this?

**Comment by dpet23** [comment]
I was thinking of writing a similar plugin when I came across this pull request, but I'm wondering why this plugin syncs only saved searches and not playlists? (I currently have several large playlists that I'd like to sync, made up of manually-curated songs with many different tags; these playlists would be very hard to replicate using saved searches.)

**Comment by gevhaz** [comment]
@daniel, there's an easy solution! One of the great things with Quod Libet is the advanced search functionality and you can do e.g. `~playlists=<playlist name>`.

File: quodlibet/quodlibet/ext/events/synchronize_to_device.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+                                        "exists.".format(dest_file))
+                                else:
+                                    append(
+                                        "Writing '{}'...".format(dest_file))
+                                    copyfile(song_path, dest_file)
+
+                            # delete files which are not
+                            # in the saved searches anymore
+                            for existing_file in Path(destination) \
+                                    .rglob('*.' + file_ext):
```
Comment (dpet23) [inline quodlibet/quodlibet/ext/events/synchronize_to_device.py:130]: If `selected_songs` is empty, `file_ext` will not be defined at this point:

```
Traceback (most recent call last):

  File "quodlibet/quodlibet/ext/events/synchronize_to_device.py", line 168, in start
    synchronize()

  File "quodlibet/quodlibet/ext/events/synchronize_to_device.py", line 134, in synchronize
    .rglob('*.' + file_ext):

UnboundLocalError: local variable 'file_ext' referenced before assignment
```

File: quodlibet/quodlibet/ext/events/synchronize_to_device.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+    config_path_key = __name__ + '_path'
+
+    def PluginPreferences(self, parent):
+        vbox = Gtk.VBox(spacing=6)
+
+        query_path = os.path.join(get_user_dir(), 'lists', 'queries.saved')
+        try:
+            with open(query_path, 'r', encoding="utf-8") as query_file:
+                if not query_file.read(1):
+                    raise FileNotFoundError
```
Comment (dpet23) [inline quodlibet/quodlibet/ext/events/synchronize_to_device.py:42]: This check prevents reading the first character of the `query_file`, and the query at the top of the file will be parsed incorrectly.

Example: with the current code, a query of `title=Sun` at the top of the file will be parsed as `itle=Sun`.

To show the "No saved searches yet" message when the file is empty, I suggest the following:

1. Remove this `if` statement:
    ```suggestion
    ```

1. _(Optional)_ Create a custom exception class to clarify why we're raising an exception:

    ```python
    class NoSavedQueriesError(Exception):
        pass
    ```

1. Add an `else` clause after reading the `query_file`:

    ```python
    queries = {}
    for query_string in query_file:
        name = next(query_file).strip()
        queries[name] = Query(query_string.strip())
    else:
        # query_file is empty
        raise NoSavedQueriesError
    ```

    or add a new `if` statement after the `for` loop:

    ```python
    if not queries:
        raise NoSavedQueriesError
    ```

1. _(Optional)_ If using the custom exception class, update the `except` clause at the bottom of the file:

    ```python
    except NoSavedQueriesError:
    ```

**Comment by Tamriel** [comment]
I fixed dpet23s bug and pulled from master to be able to merge.
Somehow the test `tests.test_util_copool.Tcopool testMethod=test_pause_restart_pause` fails. It does not fail on my machine and did not fail without the upstream changes.
Any ideas?

**Comment by Tamriel** [comment]
I think it is ready to be merged :-)

**Comment by Meriipu** [comment]
This is not really a review, but maybe consider rebasing/squashing some of the commits to make the git log cleaner and more useful (93 commits spread over time is a lot to deal with when viewing the log).

**Comment by dpet23** [comment]
> consider rebasing/squashing some of the commits to make the git log cleaner and more useful

Is it acceptable to force push after changing history? It's usually discouraged.

Also, I'd recommend reviewing the overall changes instead of commit-by-commit, since the code has been refactored several times.

**Comment by Meriipu** [comment]
> > consider rebasing/squashing some of the commits to make the git log cleaner and more useful
> 
> Is it acceptable to force push after changing history? It's usually discouraged.
>
It should not matter as long as it is an unmerged branch that is having its history edited (before being merged), since nobody is building on top of those commits yet. 
It is not really for the benefit of the reviewer as much as it is for the benefit of anybody reading the commit log in the future after the changes have been merged. 

> Also, I'd recommend reviewing the overall changes instead of commit-by-commit, since the code has been refactored several times.
>
Sure, the overall changes are easy to see when they are presented in a pull request in a combined manner.
But after they are merged, all the commits are loose again. If somebody were to view the log of quodlibet:master it would be full of these `fix merge error`-commits that do not really change anything (and there seems to be a lot of them here)

I see the value in preserving authorship in case a pull-request has several authors, and it should be possible to preserve authorship when squashing/combining commits from the same author. 
So I do not think squashing it all into a single commit is appropriate, but at least squashing the related commits together might be beneficial for future readability. 

These are just my personal opinions on it though, and I think quodlibet is/has been fairly lax on tidiness in the git log (at least for smaller pull requests with <5 commits). 
93 where a majority of them are very small changes like formatting are a bit on the excessive side in my opinion though.

**Comment by declension** [comment]
@Tamriel @Meriipu cleaning up old PRs. Is this one rescuable? If so let's try to fix the branch and any urgent review comments and get it merged (I can help perhaps)

If not let's close

**Comment by dpet23** [comment]
@declension @Meriipu, I'll squash some of the commits to make the log cleaner and (force-)push my changes in the next few days.

I think the review comments above have been addressed, but it would be useful for this PR to be reviewed again, since I've changed most of the initial implementation.

Otherwise, I think this is ready to be merged.

**Comment by Tamriel** [comment]
I squashed the commits into one. I would be happy for a final review and a merge then :-)

**Comment by declension** [comment]
Thanks!
